### ISSUE TITLE
Mask lane splits with isSkipToSpeak()=true, or else we delay the "real" turn info for too long

### ISSUE THREAD

**Issue by sonora** [osmandapp/Osmand#2571]
The carry-over from Issue #2569 is that we have "turns" in our list of route directions (in "route details" and in the left widget always displayed as "go straight") which are obsolete. They should neither be voice-announced (we have no good announcement for lane selection, and mostly there is not enough time anyway) nor displayed as a (straight) turn arrow in the left widget. (But these "turns" are be needed for our lanes widget, so they must simply be masked by isSkipToSpeak()=true.). These "turns" happen at the beginning of some divided highways where lanes split:

For where these turns are located, we currently display a "go straight" arrow (regardless whether you really need to get in the through or actually in a turn lane), so "go straight" is both obsolete and sometimes even wrong, hence this is a bug.

As these lane splits are treated as turn instructions, they "divide" the current route segment, i.e. what would normally be one segment all the way to the next "real turn" is subdivided into one segment before (showing as "go straight for x meters") and after ("in y meters turn left/right"), while the correct navigation info would all the way only be "turn left/right in x+y meters", (ignoring the interim "go straight").

**Comment by vshcherb** [comment]
Do you say that voice over says wrong distance and it is not the same as we see on the left widget.

**Comment by sonora** [comment]
No, what I am saying is this: We had always fully announced these turns (!) in all our playPrepareTurn, playMakeTurnIn, and playMakeTurn, prompts for many years, until I decided to research this now and managed to fix, see #2569. (A decisive fix worth pushing to a release some time.)

What remains is that if the next turn is one of these "hidden" turns, the navigation widget will show a "go straight" arrow (with the correct distance to the hidden turn), and tapping it produces silence (because we do not announce anything for going straight under most circumstances). Only after you passed the hidden turn will the arrow change to now display the next "real" turn, and the correct distance to it, and prompts would correctly announce the next turn and its distance. 

So now we are really only talking if it significant somewhere that on these occasions our routing shows segments of "going straight" with distances to points where no actual (real) turns happen, and only when you get there the routing indication changes to show the next real turn.

**Comment by sonora** [comment]
PS: Fixing this is not too tough, we essentially need to systematically mark or mask these points and introduce distance values relative to what matters depending on whether the nav widget, the route details screen, or the lanes widget queries the data. But these changes are rather risky, making a small mistake may mess up our routing which otherwise works real good now, and this requires lots of testing before knowing that we are back to where we were before, which we could use on more pressing issues. 

I more wanted to document this here should we ever find tis is a problem somewhere.

**Comment by sonora** [comment]
PS2: Looks like we already have some masking in place, and it is done here: https://github.com/osmandapp/Osmand/blob/master/OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java#L816. I guess the "segment interruptions" I see may be due to the ref+name change condition in that line (and not where the suppressed turns are, although this often coincides). This is fine by me for now - see no immediate need to change this.

**Comment by sonora** [comment]
Barrocas found a situation affected by this issue: It is motorway intersection http://www.openstreetmap.org/way/379672276, coming from the north, turning towards the east.

There are to consecutive exists, you have to take the second one. The first one is labeled by OsmAnd routing as a "ghost turn" via isSkipToSpeak(). Let me explain how our routing would look and sound with and without better suppression of the ghost turn:

If we managed to entirely suppress the ghost turn in our routing table, we would simply get the usual sequence of turn prompts, only referencing actual turns (and no "straights"):
 - "After 500 m turn right onto A27."  (widget arrow yellow/right, 500 m)
 - "In 300 m turn right onto A 27." (widget arrow yellow/right, 300 m)
 - "Turn right" (widget arrow green/right, ~100 m)

As we have i for now, with the ghost turn creating the extra "straight" routing segment:
 - "Continue for 500 m to A 27." (widget arrow yellow/straight, 500 m)
 - "Go straight ahead, then in 150m turn right." (widget arrow green/straight ~100m) (the 100 m is to the exit you are not taking but where you go straight)
 - (When passing the ghost turn the widget arrow changes to yellow/right, ~150 m)
 - "Turn right onto A 27." (widget arrow green/right, ~100 m)

Notice how the indications and voice prompts are a little more "crowded" in the second (our current) scenario. But Barrocas will test "in the real world", and I can imagine that at least the additional "Go straight ahead, then in 150m turn right." prompt may not even be so unwelcome by the driver, indicating you will momentarily pass an exit which _not_ to take. Hence this may be more of a feature than a bug, even if it causes the prompts being rather "dense".

**Comment by vshcherb** [comment]
1. We should not suppress turn lanes indication, it is a feature !
2. We should not pronounce ghost turn lanes where you don't make any action 
3. We should pronounce correct distances to the turn (go straight 1km and then turn right) and it should be in sync with left big widget.

I was trying to read through the code and your message but didn't find if we break any statement I made above.

**Comment by sonora** [comment]
Yes, exactly. After the recent changes I had made, the voice router does now exactly behave like you say in (2) and (3). (There had been major bugs before).

This issue is only to make us aware of what you touch in (1): Turn lanes are not only indicated in the turn lanes widget, but in these situations (exactly before a lane split with a ghost turn) also as a "straight" arrow in the left widget.

The only negative consequence of this is that if your "real" turn happens very shortly after a ghost turn, you wiil see a straight arrow saying "straight" all the way until you pass the ghost turn, and our router's preparation to indicate the following "real" turn (both in the big left arrow and by voice prompt) will as a consequence be rather short. This situation normally only happens where a ghost turn is at the beginning of some turn lanes you have to take. Big arrow indication would then be "straight" all the way until you reach the beginning of the ghost turn, only then woould it switch to "left" or "right". An example is making a left on http://www.openstreetmap.org/way/30405356 (approaching from the southeast). (Observe that there is always a "straight" ghost turn when you go straight on that same way. It is visible in the big arrow widget as a straight, but when you make no turns shortly after there is no negative impact.)

**Comment by sonora** [comment]
PS.Please note that the ghost turn is at the beginning of the turn lanes, not at the intersection itself.

**Comment by sonora** [comment]
In summary: This issue is only asking if these ghost turns should really make it to the list of our routing instructions, and hence to the left widget (there creating a straight arrow to exactly the point of the ghost turn), or if we should not confine them to the lane indication widget.

(Whatever we do, the voice router would behave accordingly and would need no changes, it prompts precisely what we show in the left widget as a result of parsing our routing instruction table.)

**Comment by Barrocas** [comment]
In that A28-A27 motorway, it works as you expected, indicating to follow
the road, go straight (with spoken destination to Viana do Castelo,
Meadela, A27), and then the turn to A27 (destination A27 Ponte de Lima).

It's correct, since it clearly indicates to go straight (not to taking the
first turn), but it seems a bit out of line from the rest os the usual
announcements, as you predicted.


-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by sonora** [comment]
Victor, based on Barrocas' and my test results, it looks like we could leave it as it is now.

But being a perfectionist, I think we could try to fix this: The underlying issue comes from RouteResultPreparation.java (either in line 899 or 926), where in some of these lane spilt cases we seem to assign "rs.speak" as 'true' to situations where there is really no speakable or actionable turn at all (mere lane splits). All other phenomena I described in the above issue result from that, I think.

**Comment by sonora** [comment]
OK, I have fixed a small bug but decisive bug we had which had caused most occurrences of these "ghost turns" (including the example I had given above): https://github.com/osmandapp/Osmand/commit/8df4dfeb6676356dd7319c4db961dea333c15843

The example Barrocas gave is actually unchanged. The routing segment change there results from the way that location is mapped: highway=motorway A 28 ends, and both ways continuing onward are mapped as highway=motorway_link. The way our code is written, and this probably makes sense in the vast majority of situations, is that we start a new routing segment when moving onto a "_link" way. I don't want to change that, so I think this issue can be called fixed.

**Comment by sonora** [comment]
I decided to back this out again. While it worked well for lane splits where you have no action (or if you have, the regular prompts for the subsequent turn would tell you), this also suppressed some "slight" turn prompts where country 2 roads simply split and you would need a prompt.

A better fix would take deeper analysis how to distinguish these 2 cases.

**Comment by sonora** [comment]
After a lot of test-driving "real world" situations I rate this is as a bug, because it has adverse effect on the quality and timing of our turn instructions (which is much better at turns not affected by this issue).

I have re-worked issue description on the very top of this with the findings and reopen this.

**Comment by vshcherb** [comment]
I still don't see the issue:
<< They should neither be voice-announced (we have no good announcement for lane selection, and mostly there is not enough time anyway) nor displayed as a (straight) turn arrow in the left widget. 

I'm not sure I know what is the issue, these turns are displayed in route info but they are not used in navigation.
https://www.dropbox.com/s/acvrffl2w5tu2rf/2016-05-25%2015.09.43.png?dl=0
This situation is displayed and  pronounced correctly, no interim announcements and distance is correct.

**Comment by vshcherb** [comment]
I'm still not sure if I understand the issue correctly but I guess it should be reproducible in my case of 60 km trip (Den Haag - Ams) but I didn't find anything strange.

**Comment by sonora** [comment]
Victor, it seems to depend on the "mapping culture", I seem to observe this more often driving in the United States or in Germany than elsewhere. I now think there may be a connection to Issue #2379, and we may find a fix for #2379 that fixes this issue as well.

If you want to reproduce it by simulation, I have researched a situation for you: Simulate a route traversing  http://www.openstreetmap.org/way/30405356, coming from the east (say 1 km east of that), with a destination so you make a left turn after that way. (Any point on http://www.openstreetmap.org/way/4900739 will work as destination.)

Now watch the left widget arrow (and listen to the turn prompts). The issue is that even though you have to make a left (which for situations not affected by the bug would have been long visible by a "Turn left" arrow in the big left widget), in this case where the bug occurs the large arrow widget will indicate "straight" all the way until you are actually where the turn lane starts. So only where the turn lane you need to take is already on your side(!), OsmAnd's display will finally indicate "Turn left" (Voice prompt is naturally delayed even a bit more). Needless to say this is usually too late, and users wonder why Osmand indicated "go straight" until much less than 65 meters away from where they have to turn left.

Side note: The bug we need to fix is actually limited when you have to make a turn in these very situations. If on the other hand you simulate a route going straight through http://www.openstreetmap.org/way/30405356, we also indicate a "go straight" turn at that point where the turn lanes start. But then it is not incorrect, and all it does is it delays the next "real" turn being indicated until after the intersection (or more precisely: after the beginning of its turn lanes) has been passed.  This is no problem if the next "real" turn is say 500m away, but is a problem if the next turn is at the very intersection only say 50 meters away.

In summary: We indicate a turn  ("go straight") at node http://www.openstreetmap.org/node/2432475847, while we fail to realize that the actual "important" turn is a "Turn left" at http://www.openstreetmap.org/node/18038479, because we do not "associate" the 2. Very similar to what osMatt describes in #2379.

**Comment by sonora** [comment]
PS: I have now simulated left turns at the 2 locations in Dallas and San Jose which osMatt had discovered to also suffer from issue #2579, and they both do NOT exhibit this #2571 bug. So the 2 issues may not be closely related after all. The difference in the example situation I specify above could be that an additional fork (not turn lane) to the right is connected to the same node where the 2 left turn lanes begin.

**Comment by sonora** [comment]
From more testing I think my "fork" assumption is correct: We do assign a "go straight" turn type at these locations which is to indicate that you do not have to take the slight right turn onto the fork. But at the same time we fail to assign the "turn left" here, because it only happens at the next (or next next) node a short distance later, while at this node only its turn lane begins. So the issue is that we discover 2 "correct" consecutive turns, first one straight, second is a left. In the mapping they happen at 2 different nodes, but for the driver in the real world it is of course the very same road intersection. We somehow need to find a way to either suppress the meaningless straight turn in these cases, or to combine all turns belonging to one intersection to the one which matters.

**Comment by sonora** [comment]
Maybe the solution is as simple as this: We assign isSkipToSpeak()=true to a TurnType.C if there is a subsequent turn within say 80 meters.

**Comment by sonora** [comment]
I have changed the title of our issue because this is really a matter of doing a better job analyzing the overall turn situations at intersections.

Here is an example of an actually rather small intersection: Take a left on http://www.openstreetmap.org/way/253048581. Also there we split it into 2 parts "Turn slightly left" (actually bogus here, if anything it is straight) where the not needed right fork is connected at node http://www.openstreetmap.org/node/2590331696, and then the real "Turn left" at node http://www.openstreetmap.org/node/1578645458, only 30 meters further.

**Comment by vshcherb** [comment]
Now I see, basically we would like to mark some "Go ahead" set pronounce false, basically because they don't indicate road split! But they indicate a turn right.
My opinion this is impossible to do without turn:lanes we should not even try that! Though if we have turn:lanes and we see right|right (so it is not through or slight_turn) we could try to skip that turn.

Though I know where right|right is marked as motorway road split and it is very advisable to speak however if it is skipped it won't be critical.

I really advise to start collecting tests very similar how we did with turn lanes and then we could really change something verifying our idea with tests. Without tests it will require too much effort to verify this discussible issue.

**Comment by sonora** [comment]
Yes I agree, it seems hard to do. The task is to distinguish

- big (cloverleaf type) motorway/highway intersections with **subsequent road splits**, where intermediate "straight" indications do make sense,

- from "normal" road intersections conceived by the driver as one and the same turn situation, where the right is just an **"integrated part of the intersection"** not to be indicated separately. (Particularly not if you need to actually make a left turn at the very same intersection.)

Here are some test cases:

(a) Small intersection: On http://www.openstreetmap.org/way/281502589, come from the south, take a left onto http://www.openstreetmap.org/way/122528745. The "straight" at node http://www.openstreetmap.org/node/986336825 is part of the very same intersection and must be ignored.

(b) Same case, slightly larger intersection: Come from the east on http://www.openstreetmap.org/way/85355911, take a left onto http://www.openstreetmap.org/way/30751813. Ignore the bogus "straight" at http://www.openstreetmap.org/node/2432475847, for the driver it belongs to the very same intersection.

(c) But here is a motorway situation:
Come from the north on http://www.openstreetmap.org/way/23441067, you want to go to http://www.openstreetmap.org/way/45039016. There is a "slight right" at http://www.openstreetmap.org/node/24836651, then a **"true straight"** at http://www.openstreetmap.org/node/253902628, (finally again a slight right at http://www.openstreetmap.org/node/17305510).

Indicating what I call the **"true straight"** in the middle is helpful for the driver, as this is a separate turn situation from the subsequent "slight right" some 400 or so meters further.

**Comment by vshcherb** [comment]
My favorite example is this http://www.openstreetmap.org/way/7364493. Here the navigation **must** provide a voice indication. You would never guess exactly right from the map configuration you gave almost the same configuration where you think it shouldn't. 

I definitely *do not recommend* to change "chatting behavior" to 'mute'. First of all first error is much less an error than second one. Second, it is a matter *user understand algorithm*. If user understand algorithm he get used to a program much faster and can react on his own, with a complex solution it will take time to double check what exactly program means and if it is right or not because it is impossible we will be 100% correct with map source that we don't control!

P.S. please don't forget that we properly ignore split of primary -> [primary, primary_link].

**Comment by sonora** [comment]
Ok, but 2 things to note here;

(1) We do never provide a standalone voice prompt to "go straight" (though we show the straight arrow). And rightly so, a repetitive voice prompt at every intersection to "go straight" would soon become annoying. We ONLY pronounce "go straight" in combination with another "real" turn imminent within  < TURN_IN_DISTANCE. (), or in combination with "...and arrive at your destination".

Hence our code slightly "misnames" what is going on: By saying "toSpeak = true/false" we mainly control if a turn is considered a separate routing instruction (which may delay other routing instructions until it has passed, but is still often not spoken if it is turnType,C), or if it is entirely ignored in the table of our routing instructions.

(2) It is true that for "Y" road splits like you mention we need to provide useful turn instructions. But mostly not "go straight", this would often be ambiguous, but either ""keep/turn:slightly right" or "keep/turn slightly left".

I think the solution could be: We "correlate" left/right turns which need to occur with any **previous** "go straights" we detected: If this combination occurs  (in this order) within say within 70m, we ignore the preceding "go straight" via isSpeak=false, hence assume it is likely part of the same intersection.

This would
- keep the (unspoken, but displayed) "straight" instruction in the motorway intersection case (400m between the straight and the subsequent turn),
- but combine (suppress) the "straight" with the subsequent left <70m away on smaller intersections?

**Comment by polarbearing** [comment]
It might be useful to speak 'straight' when the road changes name, and particularly useful when the road changes classification. It is very easy to miss a minor road you have to go straight into, because the main road makes a bend at this junction.

**Comment by vshcherb** [comment]
@sonora  Can you please send an example with subsequent "go straight" that we need to combine? Talking about (a) example, I don't understand how you could solve it. In my vision it could be only "go ahead" / "keep left" and then "turn left". I don't see how can we avoid first one without breaking other cases.

**Comment by vshcherb** [comment]
I agree that (a) and (b) quite annoying though (c) I think is perfectly ok situation. I'm thinking that we can ignore (a) and (b) only in case if there is a special lane before split, so if you drive on the right most lane before the intersection and you don't change lane just before split that it might be ok to skip voice prompt. 
In case the right most lane or even more lanes goes right  (basically after split number of lanes < before split) then we definitely need to prompt "Keep left" or "Go ahead" (the voice is nuance) but the juice is about if I need to make any action or not.

**Comment by sonora** [comment]
@vshcherb : Victor, I never talk about combining subsequent "go straights", I am solely talking about **preceding** "go straights".

If you look at the satellite pictures for routes (a) and (b) you will see that (no matter on which lane you drive) the driver would ever expect a direction to "go straight" here anywhere, expectation is to solely see (and hear) that you need to "turn left" at these intersections (a) and (b).

So our challenge is to "ignore" the preceding "go straights" at the very beginning of these intersections. And we need to do this without breaking case (c), where our go straight indication is useful and correct.

My current thinking is we can simply do this based on the distance between any "go straight" and any **subsequent** "real" turn left or right: If the distance is <=70m, we assume it belongs to the same intersection: Let is then ignore this "go straight" (in RouteResultrearation label it isSkipToSpeak(true) ). If , on the other hand, the distance is >70m (like in route (c) , we assume this is a separate turn situation, and leave everything as is. From my testing thus far this should solve all known situations.

Again: I am only talking about occurrences of a "go straight" and a subsequent "real turn left/right" **in this order**, within <~70 meters. All other situations need to remain unchanged.

**Comment by vshcherb** [comment]
I didn't talk about combining either. Personally I have no issue to set isSkipToSpeak(true) if that a discussion we have. Mostly because I don't hear voice guidance I watch which lanes to take at what time and it is impossible to hear today anyway. Though you should not put isSkipToSpeak for motorways.

**Comment by sonora** [comment]
See https://github.com/osmandapp/Osmand/commit/168ddd955bff82b2e17b7f65e3b5218a2147fdd1. This solves all our known test cases, and keeps full functionality on motorway intersections.

Funny enough: While the unwanted goStraights seem reliably suppressed in the left widget stack and for the voice prompts, they are under some circumstances not hidden on the Route Details Info screen (then when you move beyond certain turns they are). I believe this may be a timing or update issue of the list adapter on that screen. But this issue seems reliably solved, regardless.

**Comment by vshcherb** [comment]
Good job, I already noticed the positive effect while was fixining related issue with left turn on turn lanes indicator.
